lib/projfs.c

```diff
@@ -144,88 +144,89 @@ static int projfs_fuse_perm_event(uint64_t mask,
 
 struct node_userdata
 {
+	int fd;
+
 	uint8_t proj_flag; // USER_PROJECTION_EMPTY
 };
 
 #define USER_PROJECTION_EMPTY "user.projection.empty"
 
-static __thread char *mapped_path;
-
-static void set_mapped_path(char const *path, int parent)
+/**
+ * If parent is 0, returns a copy of path.
+ *
+ * If parent is non-zero, return a copy of path with the last component removed
+ * (e.g. "x/y/z" will yield "x/y").  If path has only one component,
+ * returns ".".
+ *
+ * The caller is responsible for freeing the returned string.
+ */
+static char *get_mapped_path(char const *path, int parent)
 {
 	const char *last;
 
-	free(mapped_path);
-	if (!parent) {
-		mapped_path = strdup(path);
-		return;
-	}
+	if (!parent)
+		return strdup(path);
+
 	last = strrchr(path, '/');
 	if (!last)
-		mapped_path = strdup(".");
+		return strdup(".");
 	else
-		mapped_path = strndup(path, last - path);
+		return strndup(path, last - path);
 }
 
 /**
- * @return the userdata struct pointer, or NULL and sets errno.  If non-NULL is
- *         returned, the FUSE userdata is acquired and must be released with
- *         fuse_context_node_userdata_release.
+ * Acquires a lock on mapped_path and populates the supplied node_userdata
+ * argument with the open and locked fd, and proj_flag based on the
+ * USER_PROJECTION_EMPTY xattr.
+ *
+ * @return 0 or an errno
  */
-static struct node_userdata *get_path_userdata(int parent)
+static int get_path_userdata(struct node_userdata *user, const char *mapped_path)
 {
-	struct timespec abs_timeout;
-	struct fuse_node_userdata *fuserdata;
-	struct node_userdata *user;
 	ssize_t sz;
-	int fd, err;
+	int err;
 
-	clock_gettime(CLOCK_REALTIME, &abs_timeout);
-	abs_timeout.tv_sec += PROJ_WAIT_SEC;
-	abs_timeout.tv_nsec += PROJ_WAIT_NSEC;
+	bzero(user, sizeof(*user));
```

Comment:
I might suggest `memset()` here as `bzero()` is deprecated or removed these days (see [bzero(3)](http://man7.org/linux/man-pages/man3/bzero.3.html#CONFORMING_TO)).